### PR TITLE
Position FSM

### PR DESCRIPTION
**Pull request by rishiso**
## Description
Implements position management automatically based on states of the game. Positions used to be static (3 defense, 2 offense, 1 goalie), but now they change dynamically.

## Associated / Resolved Issue
[ClickUp card](https://app.clickup.com/t/86aypjcg3)

## Steps to Test
Run in sim and move the ball to different halves. The positions of robots should change dynamically.

## Key Files to Review
 * robot_factory_postion.cpp

## Review Checklist

- [ ] **Docstrings**: All methods and classes should have the file appropriate docstrings which follow the guidelines in the ["Contributing" page](https://rj-rc-software.readthedocs.io/en/latest/contributing.html) of our docs.
- [ ] **Remove extra print statements**: Any print statements used for debugging should be removed
- [ ] **Tag reviewers**: Tag some people for review and ping them on Slack

### CODE REVIEW

File: soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp [hunk 1]
```diff
@@ -16,8 +16,63 @@ RobotFactoryPosition::RobotFactoryPosition(int r_id) : Position(r_id) {
 
 std::optional<RobotIntent> RobotFactoryPosition::get_task(WorldState& world_state,
                                                           FieldDimensions& field_dimensions) {
-    // This is where the current_position can be reassigned based on the
-    // PlayState
+    // If keeper, make no changes
+    if (robot_id_ == 0) {
```
Comment (sid-parikh) [inline soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp:20]: you can't assume 0 is the robot id. you need to get that information from the referee. If i'm not mistaken, there is already a mechanism for AAC to pass this information to its Position. if not, let's figure out how to add it.

File: soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp [hunk 2]
```diff
@@ -16,8 +16,63 @@ RobotFactoryPosition::RobotFactoryPosition(int r_id) : Position(r_id) {
...
+        return current_position_->get_task(world_state, field_dimensions);
+    }
+
+    // TODO (Rishi and Jack): Make this synchronized across all robots to avoid race conditions
+
+    // Get sorted positions of all friendly robots
+    typedef struct std::pair<int, double> RobotPos;  // (robotId, yPosition)
```
Comment (sid-parikh) [inline soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp]:
```suggestion
    using RobotPos = std::pair<int, double>;  // (robotId, yPosition)
```
note the GH Action warning here. just simpler syntax it's completely equivalent

File: soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp [hunk 3]
```diff
@@ -16,8 +16,63 @@ RobotFactoryPosition::RobotFactoryPosition(int r_id) : Position(r_id) {
...
+
+    std::vector<RobotPos> robots_copy;
+    for (int i = 0; i <= 5; i++) {
```
Comment (sid-parikh) [inline soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp]: you also shouldn't assume our robots are only 0 through 5; use `kNumShells`, or, better yet, just use  `world_state.our_robots.size()`

File: soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp [hunk 4]
```diff
@@ -16,8 +16,63 @@ RobotFactoryPosition::RobotFactoryPosition(int r_id) : Position(r_id) {
...
+        if (r.first == robot_id_) {
+            break;
+        }
+        i++;
+    }
+
+    // Assigning new position
+    // Checking whether we have possesion or if the ball is on their half (using 1.99 to avoid
+    // undetermined behavior on midline)
+    if (Position::our_possession_ ||
```
Comment (sid-parikh) [inline soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp:53]: what's the undetermined behavior?

**Review by sid-parikh — CHANGES_REQUESTED** [review]

File: soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp (hunk 4 above)
Comment (rishiso) [reply soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp:53]: I noticed that the robots sometimes switched rapidly between offense and defense (probably due to rounding).

**Review by sid-parikh — APPROVED** [review]
hey don't worry about the Goalie fix for now; I worked on that for the Offense PR anyway

**Comment by sid-parikh** [comment]
your last commit says (not finished) so I'm not hitting merge, but feel free to push it if you're done.

**Pull request merged**